src/types/mmapped.jl

```diff
@@ -0,0 +1,76 @@
+struct MmappedTIFF{T <: Colorant, N, O <: Unsigned, A <: AbstractMatrix{T}} <: AbstractTIFF{T, N}
+    """
+    2d slices in the file"
+    """
+    chunks::Vector{A}
+    """
+    The associated tags for each slice in this array
+    """
+    ifds::Vector{IFD{O}}
+    """
+    The 2d slice size
+    """
+    sz2::Dims{2}
+end
+
+function MmappedTIFF{T,N}(file::TiffFile{O}, ifds::Vector{IFD{O}}) where {T, N, O <: Unsigned}
+    N ∈ (2, 3) || error("only 2d and 3d TIFF arrays are supported")
+    if N == 2
+        length(ifds) == 1 || error("2d files must have a single Image File Directory")
+    end
+    ifd = first(ifds)
+    sz = (nrows(ifd), ncols(ifd))
+    compression = getdata(CompressionType, ifd, COMPRESSION, COMPRESSION_NONE)
+    compression == COMPRESSION_NONE || error("mmap is not yet supported with compression")
+    all(iscontiguous, ifds) || error("non-contiguous TIFFs are not yet supported by mmap")
+
+    seek(file, 0)
+    rawio = stream(file)
+    raw = Mmap.mmap(rawio, Vector{UInt8}, filesize(rawio))
+    chunks = [getchunk(T, raw, reverse(sz), ifd) for ifd in ifds]
+    return MmappedTIFF{T, N, O, eltype(chunks)}(chunks, ifds, sz)
+end
+
+function MmappedTIFF{T}(file::TiffFile{O}, ifds::Vector{IFD{O}}) where {T, O <: Unsigned}
+    l = length(ifds)
+    return l == 1 ? MmappedTIFF{T, 2}(file, ifds) : MmappedTIFF{T, 3}(file, ifds)
+end
+
+function MmappedTIFF(file::TiffFile{O}, ifds::Vector{IFD{O}}) where {O <: Unsigned}
+    ifd = first(ifds)
+    T = rawtype(ifd)
+    colortype, _ = interpretation(ifd)
+    return MmappedTIFF{colortype{_mappedtype(T)}}(file, ifds)
+end
+
+function getchunk(::Type{T}, raw::Vector{UInt8}, sz::Dims{2}, ifd::IFD) where T
+    rawsz = sizeof(T) * prod(sz)
+    strip_offsets = ifd[STRIPOFFSETS].data
+    o = strip_offsets[1]::Core.BuiltinInts
+    if sizeof(T) == 1
+        return reinterpret(reshape, T, reshape(view(raw, o+1:o+rawsz), sz...))
+    end
+    return reinterpret(reshape, T, reshape(view(raw, o+1:o+rawsz), sizeof(T), sz...))
+end
+
+## AbstractArray interface
+
+chunk1(img::MmappedTIFF) = @inbounds return img.chunks[1]
+
+Base.size(img::MmappedTIFF{T, 2}) where T = img.sz2
+Base.size(img::MmappedTIFF{T, 3}) where T = (img.sz2..., length(img.chunks))
+
+Base.@propagate_inbounds Base.getindex(img::MmappedTIFF{T, 2}, i::Int, j::Int) where T = chunk1(img)[j, i]
+
+Base.@propagate_inbounds function Base.getindex(img::MmappedTIFF{T, 3}, i::Int, j::Int, k::Int) where T
+    chunk = img.chunks[k]
+    return chunk[j, i]
+end
+
+# These work only if the file was opened with write permissions
+Base.@propagate_inbounds Base.setindex!(img::MmappedTIFF{T, 2}, val, i::Int, j::Int) where T = chunk1(img)[j, i] = val
+
+Base.@propagate_inbounds function Base.setindex!(img::MmappedTIFF{T, 3}, val, i::Int, j::Int, k::Int) where T
+    chunk = img.chunks[k]
+    chunk[j, i] = val
+end
```

Comment:
```julia
julia> img = TiffImages.load("tmp.tif"; mmap=true)

julia> img[2] = img[1]
ERROR: ReadOnlyMemoryError()
Stacktrace:
  [1] setindex!
    @ .\array.jl:903 [inlined]
  [2] setindex!
```

An idea here can be explicitly constructing the `MmappedTIFF` with a readonly bool value so that we can eagerly detect and throw errors.

```julia
struct MmappedTIFF{T <: Colorant, N, Writable, O <: Unsigned, A <: AbstractMatrix{T}} <: AbstractTIFF{T, N}
...
end
...

Base.@propagate_inbounds Base.setindex!(img::MmappedTIFF{T, 2, true}, val, i::Int, j::Int) where T = chunk1(img)[j, i] = val
Base.@propagate_inbounds function Base.setindex!(img::MmappedTIFF{T, 3}, val, i::Int, j::Int, k::Int) where T
    chunk = img.chunks[k]
    chunk[j, i] = val
end
function Base.setindex!(img::MmappedTIFF{T, N, false}, val, i::Int, j::Int) where {T,N}
    error("unable to write a readonly memory-mapped array, maybe you need to open the file with `r+` mode.")
end
```

Reply:
That's a good idea. Just to decrease need for compilation specialization (some may use both read-only and read-write in the same session), do you think we could do almost as well with `register_error_hint`? If so, that would require less duplication of compilation effort.

Reply:
The duplication of compilation is a valid reason to not do this.

But I'm not sure whether we can register an error hint for `ReadOnlyMemoryError` because [it's an empty struct](https://github.com/JuliaLang/julia/blob/6ce28008ba6db324b171909fa8e641fe8bce9db4/base/boot.jl#L282-L287) that we have no extra information
to examine and filter. Can `register_error_hint` get callstack information in general?

If not doable, the current best approach is just to mention `ReadOnlyMemoryError` error is due to file mode in the documentation.